### DISCUSSION TITLE
JSBSim sideslip angle, rudder kick

### DISCUSSION THREAD

**Discussion by aregis74** [JSBSim-Team/jsbsim#519] [Q&A]
Hello guys! I trying to do a rudder kick maneuver, but I didn't find how to stabilize the aircraft with an sideslip angle, somebody has a clue to execute the script?

I tried to apply do.trim(5) in turn with  ic/beta-deg as target, but I didn't succeed.

thank you for someone help!

**Comment by seanmcleod** [comment]
I don't think you can set a beta target with `tTurn = 5`, for `tTurn` the trim code varies `beta` in order to get  `Vdot` to 0.

https://github.com/JSBSim-Team/jsbsim/blob/dd97e99032e6cce6ae3d0d91730f3c10f8ce94af/src/initialization/FGTrim.cpp#L839-L846

**Reply by aregis74** [reply]
Dear Sean, I appreciate your answer and I gessed it wasn't possible. How about the mode tcustom? Is there any reference about how I can apply this mode? Or do you have any clue how can I trim the aircraft with sidelip angle?
Thank you again for your time,
Regards.

**Comment by seanmcleod** [comment]
Correct, custom is an option. Haven't looked into how you specifically setup custom though. But for example you could potentially set up a custom version that is the same as `tTurn` but just removes the `(tVdot, tBeta)` trim axis, assuming you didn't care about the resulting `Vdot`.

What exactly are your requirements in terms of `Udot, Vdot, Wdot`? Are you looking to setup for example a steady heading sideslip configuration? Or do you actually want a turn, but uncoordinated with a non-zero beta?

**Reply by aregis74** [reply]
I am just looking to setup for a steady heading sideslip.

**Comment by seanmcleod** [comment]
Okay, so you still want `Udot, Vdot, Wdot` to all be 0. But now you'll need aileron to control both `Pdot` and `Vdot`.

**Reply by aregis74** [reply]
Correct, I thought there was already a function to control the pdot and vdot for steady heading sideslip.
thank you again for your time.

**Comment by seanmcleod** [comment]
Actually, double-checking, have you tried `tFull`?

https://github.com/JSBSim-Team/jsbsim/blob/dd97e99032e6cce6ae3d0d91730f3c10f8ce94af/src/initialization/FGTrim.cpp#L805-L815

Bank angle (phi) is used to control `Vdot` and aileron for `Pdot`. So if you provide a beta value as part of your initial conditions it looks like `tFull` may be able to generate a steady heading sideslip solution?

**Comment by seanmcleod** [comment]
@aregis74 I tried a quick test using the full trim option and it seems to work in terms of finding trim solutions for steady heading sideslip.

I used the 737 model and varied the initial beta from 0 to 5 degrees and then plotted the resulting bank angle, aileron and rudder commands from the trim solutions.

![image](https://user-images.githubusercontent.com/824856/143684131-ada3afbf-d985-4eee-8489-526b996a0b62.png)

```python
import jsbsim
import matplotlib.pyplot as plt
import math

fdm = jsbsim.FGFDMExec('..\\') # The path supplied to FGFDMExec is the location of the folders "aircraft", "engines" and "systems"

fdm.load_model('737') # Load the aircraft 737

# Set engines running
fdm['propulsion/engine[0]/set-running'] = 1
fdm['propulsion/engine[1]/set-running'] = 1

# Set alpha range for trim solutions
fdm['aero/alpha-max-rad'] = math.radians(12)
fdm['aero/alpha-min-rad'] = math.radians(-4.0)


betas = []

bankAngle = []
ailerons = []
rudder = []

for beta in range(0, 6, 1):
    fdm['ic/h-sl-ft'] = 1000
    fdm['ic/vc-kts'] = 200
    fdm['ic/gamma-deg'] = 0
    fdm['ic/beta-deg'] = beta

    fdm.run_ic() # Initialize the aircraft with initial conditions

    fdm.run()

    # Trim
    try:
        fdm['simulation/do_simple_trim'] = 1
        betas.append(fdm['aero/beta-deg'])
        bankAngle.append(fdm['attitude/phi-deg'])
        ailerons.append(fdm['fcs/aileron-cmd-norm'])
        rudder.append(fdm['fcs/rudder-cmd-norm'])
    except RuntimeError as e:
        # The trim cannot succeed. Just make sure that the raised exception
        # is due to the trim failure otherwise rethrow.
        if e.args[0] != 'Trim Failed':
            raise

fig, ax1 = plt.subplots()
ax1.set_xlabel('Beta (deg)')
ax1.set_ylabel('Bank Angle (deg)')
ax1.set_ylim(0, 9)
ax1.plot(betas, bankAngle, label='Bank Angle', color='red')

ax2 = ax1.twinx()

ax2.set_ylabel('Inceptor Position')
ax2.set_ylim(0, 1)
ax2.plot(betas, ailerons, label='Aileron')
ax2.plot(betas, rudder, label='Rudder')

ax1.legend(loc=0)
ax2.legend(loc=2)

plt.title('Steady Heading Sideslip')

plt.show()

```

**Reply by aregis74** [reply]
Yeah, I've got the same results, thank you. However I actually need is to simulate a rudder kick maneuver, that means to find the ailerons positions for a maximum rudder deflection to maintain in steadyhead sideslip. The trim tfull will find the ailerons to maintain in steadyhead sideslip, but with beta as input, and I need rudder as input. Even so,  your sugestions did help a lot. 
Regards

**Comment by seanmcleod** [comment]
Okay, although your initial request was a bit misleading then, since you specifically mentioned setting a beta target.

However you could simply extend my beta loop to continue in small beta increments until the trim solution returned a solution with the rudder at maximum deflection, noting the aileron deflection and assuming that the rudder deflection reaches maximum deflection before the aileron deflection does, the beta to achieve this and the roll angle to achieve this.

**Comment by seanmcleod** [comment]
So for the 737 example we end up with the aileron deflection hitting a maximum before the rudder deflection does. At ~13.5 deg for beta we have aileron deflection at 1 and rudder deflection at 0.92, with a bank angle of ~22.5 deg.

![image](https://user-images.githubusercontent.com/824856/143876100-ce3635d9-7c2c-4005-aaab-9ab5c7c91a33.png)

**Reply by aregis74** [reply]
Sorry for the misunderstanding, I was trying to do a Rudder Kick maneuver since the first request. Setting a beta target is an option to complete it. I'll folow yor sugestion of beta increments, and I'll share my results here. I need to find a time history like this:
![image](https://user-images.githubusercontent.com/71350921/143879384-ba7d45a0-21d6-49e4-892f-2d5d4ff5c4a9.png)

**Comment by seanmcleod** [comment]
Ah, I concentrated/fixated on the comments regarding trim and setting beta. If I'd seen the diagram you just posted and/or this description of rudder kick I would've suggested something different.

> At an operationally safe altitude, and using a prominent geographic feature, establish the aircraft in straight and level flight at the maximum Vfe speed (for target purposes).  Initiate smooth and continuous rudder pressure until full deflection to the left or right is established; simultaneously, initiate roll in the opposite direction sufficient to maintain a straight track whereby the aircraft’s longitudinal axis remains in line with the targeted geographic feature.

So if you're trying to automate the rudder kick then I'd suggest the following.

Implement a PID controller using the ailerons to keep either the heading rate = 0 or Vdot = 0.

Use the full trim option to trim the aircraft for straight and level flight at the specific altitude and airspeed you want to use. The PID gains need to be optimised for this altitude and airspeed.

Then use an event to perform a ramp input for the rudder to full deflection. If the PID controller is well tuned then it should keep the aircraft on a steady heading as the rudder input kicks in.

Lastly setup output directives to log the relevant parameters you want a time history of.

If you don't want to take on the task of tuning a PID controller I guess you could potentially try the following in terms of generating a similar ramp input for the ailerons, and see whether it does a good enough job.

What I mean in terms of similar is that if you look at the slopes of the aileron and rudder inputs from my last graph, you could come up with a ramp for the aileron that ramps up slightly faster than the rudder ramp, i.e. so that it is always slightly larger than the rudder input, with a maximum value of say 1 versus a maximum value of 0.92 for the rudder input.

**Comment by seanmcleod** [comment]
Just tried my suggestion out and these are the results I get for the 737 model.

![image](https://user-images.githubusercontent.com/824856/143937791-4a9c6dda-d7ea-4da2-9781-650568ec69b3.png)

Using the following code.

```python
import jsbsim
import matplotlib.pyplot as plt
import math
import numpy as np

fdm = jsbsim.FGFDMExec('..\\') # The path supplied to FGFDMExec is the location of the folders "aircraft", "engines" and "systems"

fdm.load_model('737') # Load the aircraft 737

# Set engines running
fdm['propulsion/engine[0]/set-running'] = 1
fdm['propulsion/engine[1]/set-running'] = 1

# Set alpha range for trim solutions
fdm['aero/alpha-max-rad'] = math.radians(12)
fdm['aero/alpha-min-rad'] = math.radians(-4.0)

dt = 1/120

aileronMax = 1
rudderMax = 0.92

risetime = 1
diAileron = aileronMax / (risetime/dt)
diRudder = rudderMax / (risetime/dt)

times = []

betas = []
bankAngle = []
ailerons = []
rudder = []

fdm['ic/h-sl-ft'] = 1000
fdm['ic/vc-kts'] = 200
fdm['ic/gamma-deg'] = 0
fdm['ic/beta-deg'] = 0

fdm.run_ic() # Initialize the aircraft with initial conditions

fdm.run()

# Trim
try:
    fdm['simulation/do_simple_trim'] = 1

    for i in range(int(10/dt)):
        fdm.run()

        times.append(fdm.get_sim_time())

        betas.append(fdm['aero/beta-deg'])
        bankAngle.append(fdm['attitude/phi-deg'])
        ailerons.append(fdm['fcs/aileron-cmd-norm'])
        rudder.append(fdm['fcs/rudder-cmd-norm'])

        aileronCmd = fdm['fcs/aileron-cmd-norm']
        rudderCmd = fdm['fcs/rudder-cmd-norm']

        if aileronCmd < aileronMax:
            aileronCmd += diAileron
            fdm['fcs/aileron-cmd-norm'] = aileronCmd

        if rudderCmd < rudderMax:
            rudderCmd += diRudder
            fdm['fcs/rudder-cmd-norm'] = rudderCmd

except RuntimeError as e:
    # The trim cannot succeed. Just make sure that the raised exception
    # is due to the trim failure otherwise rethrow.
    if e.args[0] != 'Trim Failed':
        raise

fig, ax1 = plt.subplots()
ax1.set_xlabel('Time (s)')
ax1.set_ylabel('Beta (deg)')
line1 = ax1.plot(times, betas, label='Beta', color='red')

ax2 = ax1.twinx()

ax2.set_ylabel('Inceptor Position')
line2 = ax2.plot(times, ailerons, label='Aileron')
line3 = ax2.plot(times, rudder, label='Rudder')

ax1.legend(handles=line1+line2+line3, loc=4)

plt.title('Rudder Kick')

plt.show()
```

**Reply by bcoconni** [reply]
@seanmcleod I have pushed a commit to the `master` branch to avoid using the following idiom to catch an exception thrown by a trim failure
```python
except RuntimeError as e:
    # The trim cannot succeed. Just make sure that the raised exception
    # is due to the trim failure otherwise rethrow.
    if e.args[0] != 'Trim Failed':
        raise
```
Since commit a5484eeea23cb15492b9d73f7ef1e833e025a99c, the same result can be achieved with the (hopefully) more explicit syntax
```python
except jsbsim.TrimFailureError:
    pass
```

**Reply by bcoconni** [reply]
Note that the change is backward compatible and that the syntax you used above is still working.

**Reply by seanmcleod** [reply]
@bcoconni thanks. I had actually just looked at the commit a bit before your messages here. This will be clearer.

**Comment by seanmcleod** [comment]
Example of changing the rise time, from 1s to 3s for maximum inceptor deflections and showing the time history for 20s.

![image](https://user-images.githubusercontent.com/824856/143938671-46eb8317-a772-4c0a-b600-cab7cb9838f2.png)
